fifteen_min/src/find_home.rs

```diff
@@ -89,33 +89,28 @@ fn score_houses(
     options: Options,
     timer: &mut Timer,
 ) -> HashMap<BuildingID, Percent> {
-    let mut satisfied_per_bldg: Counter<BuildingID> = Counter::new();
-
-    // There may be more clever ways to calculate this, but we'll take the brute-force approach of
-    // calculating the walkshed from every matching business.
     let num_categories = amenities.len();
-    for category in amenities {
-        let mut stores: HashSet<BuildingID> = HashSet::new();
-        for b in app.map.all_buildings() {
-            if b.has_amenity(category) {
-                stores.insert(b.id);
-            }
-        }
+    let mut satisfied_per_bldg: Counter<BuildingID> = Counter::new();
 
-        let mut houses: HashSet<BuildingID> = HashSet::new();
-        let map = &app.map;
-        for times in timer.parallelize(
-            &format!("find houses close to {}", category),
-            Parallelism::Fastest,
-            stores.into_iter().collect(),
-            |b| options.clone().time_to_reach_building(map, b),
-        ) {
-            for (b, _) in times {
-                houses.insert(b);
+    let map = &app.map;
+    for times in timer.parallelize(
+        "find houses close to amenities",
+        Parallelism::Fastest,
+        amenities,
+        |category| {
+            // For each category, find all matching stores
+            let mut stores = Vec::new();
+            for b in map.all_buildings() {
```

Comment:
If you're looking for further speedups, one thing that comes to mind...

Currently we're iterating through all buildings once per amenity. 

Since amenities is relatively small and buildings is relatively large, it might be faster to do something parallelizing on buildings rather than amenities, so that we're only iterating through buildings once, rather than once per amenity (even if parallel, the speedup isn't typically 1:1). 

What's more, I don't think you need to actually iterate through amenities, assuming they're hashable you could do a hash set intersection.

Take it or leave it though, if it's already fast enough.

Reply:
Hmm, maybe I'm confused though - maybe it's that the major perf win is coming from sharing state between doing multiple pathfindings all at once?

Reply:
Iterating through all buildings in the map to look for stores should already be fast. We could make a single pass through to gather the stores matching each category, then parallelize the pathfinding. But I'm not sure it's worth the extra code to track stores per amenity. I haven't done a timing breakdown, but I'd be really surprised if `times_from_buildings` isn't the bottleneck here.

The perf win comes from reducing the number of Dijkstra floodfills from O(total number of matching stores) to O(number of amenity categories). The individual floodfills may get a little slower, because there are more starting points, so the total reachable 15m walkshed is larger. But this is a vast win over repeating a floodfill from every single store.

Reply:
I wish I could whiteboard this really quickly, but another explanation: before, for `AmenityType::Food`, we take the 200 stores with that amenity and individually compute a 15m walkshed from each of them. Then we union the results.

After, we compute just one 15m walkshed for `AmenityType::Food`. The queue of nodes to visit in Dijkstra's just starts with 200 nodes (cost 0), instead of 1.

Reply:
That all makes sense, thanks for explaining!

Reply:
Thats really cool. Its funny I always see Dijkstra starting from 1 node and always think of it that way but yes starting with 200 nodes for example and running it once makes a lot of sense, after all you don't really care about the walkshed for each individual building, you care about the union of all of them.